platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/cli/converter/LayoutToPropertiesConverter.java

```diff
@@ -56,6 +57,7 @@ public LayoutToPropertiesConverter(BuildLayoutFactory buildLayoutFactory) {
         allBuildOptions.addAll(new LoggingConfigurationBuildOptions().getAllOptions()); // TODO maybe a new converter also here
         allBuildOptions.addAll(new WelcomeMessageBuildOptions().getAllOptions()); // TODO maybe a new converter also here
         allBuildOptions.addAll(new DaemonBuildOptions().getAllOptions());
+        allBuildOptions.addAll(new DaemonJvmToolchainCriteriaOptions().getAllOptions());
```

Comment:
I'm not sure we want to mix-in the DaemonJvmToolchainCriteriaOptions with the rest of the options here. 

With this code here, that means all build properties will be permitted if they're in `gradle.properties` and all gradle properties will be permitted if they're in `gradle/gradle-build.properties`. We should have separate lists of build options for each properties files.

Reply:
Can we write tests to make sure `gradle.properties` props are not allowed in `gradle/gradle-build.properties` and vice versa?

Reply:
The idea was to reuse in the first place the existing `BuildOption` implementation for this reason was necessary to make the existing `AbstractBuildOption` generic allowing different `PropertyOrigin` the existing `gradle.properties` being this the first refactor [commit](https://github.com/gradle/gradle/pull/27898/commits/85b0ea67e3a02833f1b4506e90503855b6469d84). This commit tries to archive with the minimum number of changes to have different properties origin, for this reason in the case of `DaemonJvmToolchainCriteriaOptions` the origin is specified as `PropertyOrigin.BUILD_PROPERTIES` while by default the `PropertyOrigin.GRADLE_PROPERTIES` is used, allowing us to reuse `StringBuildOption` and `EnumBuildOption`. 

I considered this much better approach than having to duplicate them unnecessary since conceptually makes completely sense to be able to have different property origin being this the existing `gradle.properties` or the new `gradle-build.properties` or even any other in the future. For this reason, I exposed separated the build properties on `AllProperties` to avoid explicitly to have mix properties from different origin (the problem you describe). Ideally the existing `getProperties`, should be a map where the origin is also specified, however, this require a massive refactor for this reason I avoided to do it, keeping for the moment separated.
```    
public interface AllProperties {
    Map<String, String> getProperties();
}
```

Regarding tests, yes we already validate this on `DaemonToolchainCoexistWithCurrentOptionsIntegrationTest`:
 - Given daemon toolchain properties defined on gradle properties When executing any task Then those are ignored since aren't defined on build properties file
 - Given daemon toolchain properties When executing any task passing them as arguments Then those are ignored since aren't defined on build properties file

The one that is missing is when defining the `org.gradle.java.home` under for example build properties that I'm going to add to test in this way the left scenario